# TELCO PROJECT

# GOAL:
* Words go Here
* More Words go Here

# IMPORTS

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
import os

from sklearn.model_selection import train_test_split
import sklearn.preprocessing
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression

#import warnings
#warnings.filterwarnings("ignore")

from scipy import stats
import re

import wrangle as w
import QMCBT_explore_evaluate as ee
#import QMCBT_modeling as m

In [2]:
# Acquire telco_churn data from Codeup SQL database
df = w.get_telco_churn_df()

# EXPLORE

In [3]:
df.T

0                    1     \
customer_id                          0002-ORFBO           0003-MKNFE   
payment_type_id                               2                    2   
contract_type_id                              2                    1   
internet_service_type_id                      1                    1   
gender                                   Female                 Male   
senior_citizen                                0                    0   
partner                                     Yes                   No   
dependents                                  Yes                   No   
tenure                                        9                    9   
phone_service                               Yes                  Yes   
multiple_lines                               No                  Yes   
online_security                              No                   No   
online_backup                               Yes                   No   
device_protection                            No                   No   
tech_support                                Yes                   No   
streaming_tv                                Yes                   No   
streaming_movies                             No                  Yes   
paperless_billing                           Yes                   No   
monthly_charges                            65.6                 59.9   
total_charges                             593.3                542.4   
churn                                        No                   No   
internet_service_type                       DSL                  DSL   
contract_type                          One year       Month-to-month   
payment_type                       Mailed check         Mailed check   
churn_month                                 NaN                  NaN   
contract_type_id.1                            2                    1   
paperless_billing.1                         Yes                   No   
gender.1                                 Female                 Male   
senior_citizen.1                              0                    0   
partner.1                                   Yes                   No   
dependents.1                                Yes                   No   
payment_type_id.1                             2                    2   
monthly_charges.1                          65.6                 59.9   
total_charges.1                           593.3                542.4   
signup_date                 2021-04-21 18:07:34  2021-04-21 18:07:34   
phone_service.1                             Yes                  Yes   
multiple_lines.1                             No                  Yes   
internet_service_type_id.1                    1                    1   
online_security.1                            No                   No   
online_backup.1                             Yes                   No   
device_protection.1                          No                   No   
tech_support.1                              Yes                   No   
streaming_tv.1                              Yes                   No   
streaming_movies.1                           No                  Yes   

                                           2                    3     \
customer_id                          0004-TLHLJ           0011-IGKFF   
payment_type_id                               1                    1   
contract_type_id                              1                    1   
internet_service_type_id                      2                    2   
gender                                     Male                 Male   
senior_citizen                                0                    1   
partner                                      No                  Yes   
dependents                                   No                   No   
tenure                                        4                   13   
phone_service                               Yes                  Yes   
multiple_l

In [4]:
ee.explore_toc()

** CUSTOM EXPLORATION FUNCTIONS
explore_tips: PRINT A LIST OF USEFUL FUNCTIONS, METHODS, AND ATTRIBUTES USED FOR EXPLORATION
nunique_column_all(df): PRINT NUNIQUE OF ALL COLUMNS
nunique_column_objects(df): PRINT NUNIQUE OF COLUMNS THAT ARE OBJECTS
nunique_column_qty(df): PRINT NUNIQUE OF COLUMNS THAT ARE *NOT* OBJECTS
numeric_range(df): COMPUTE RANGE FOR ALL NUMERIC VARIABLES


In [5]:
ee.explore_tips()

** USEFUL EXPLORATORY CODE**
DFNAME.head()
DFNAME.shape
DFNAME.shape[0] #read row count
DFNAME.describe().T
DFNAME.columns.to_list()
DFNAME.COLUMNNAME.value_counts(dropna=False)
DFNAME.dtypes
DFNAME.select_dtypes(include='object').columns
DFNAME.select_dtypes(include='float').columns
pd.crosstab(DFNAME.COLUMN-1, DFNAME.COLUMN-2)


In [6]:
# return (Row, Column)
df.shape

(7043, 44)

In [7]:
# Return non-null count and Dtype for each column
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 44 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   customer_id                 7043 non-null   object 
 1   payment_type_id             7043 non-null   int64  
 2   contract_type_id            7043 non-null   int64  
 3   internet_service_type_id    7043 non-null   int64  
 4   gender                      7043 non-null   object 
 5   senior_citizen              7043 non-null   int64  
 6   partner                     7043 non-null   object 
 7   dependents                  7043 non-null   object 
 8   tenure                      7043 non-null   int64  
 9   phone_service               7043 non-null   object 
 10  multiple_lines              7043 non-null   object 
 11  online_security             7043 non-null   object 
 12  online_backup               7043 non-null   object 
 13  device_protection           7043 

In [8]:
df.describe().T

count       mean        std    min   25%    50%  \
payment_type_id             7043.0   2.315633   1.148907   1.00   1.0   2.00   
contract_type_id            7043.0   1.690473   0.833755   1.00   1.0   1.00   
internet_service_type_id    7043.0   1.872923   0.737796   1.00   1.0   2.00   
senior_citizen              7043.0   0.162147   0.368612   0.00   0.0   0.00   
tenure                      7043.0  32.371149  24.559481   0.00   9.0  29.00   
monthly_charges             7043.0  64.761692  30.090047  18.25  35.5  70.35   
contract_type_id.1          7043.0   1.690473   0.833755   1.00   1.0   1.00   
senior_citizen.1            7043.0   0.162147   0.368612   0.00   0.0   0.00   
payment_type_id.1           7043.0   2.315633   1.148907   1.00   1.0   2.00   
monthly_charges.1           7043.0  64.761692  30.090047  18.25  35.5  70.35   
internet_service_type_id.1  7043.0   1.872923   0.737796   1.00   1.0   2.00   

                              75%     max  
payment_type_id              3.00    4.00  
contract_type_id             2.00    3.00  
internet_service_type_id     2.00    3.00  
senior_citizen               0.00    1.00  
tenure                      55.00   72.00  
monthly_charges             89.85  118.75  
contract_type_id.1           2.00    3.00  
senior_citizen.1             0.00    1.00  
payment_type_id.1            3.00    4.00  
monthly_charges.1           89.85  118.75  
internet_service_type_id.1   2.00    3.00

In [9]:
ee.nunique_column_objects(df)

customer_id has 7043 unique values.
gender has 2 unique values.
partner has 2 unique values.
dependents has 2 unique values.
phone_service has 2 unique values.
multiple_lines has 3 unique values.
online_security has 3 unique values.
online_backup has 3 unique values.
device_protection has 3 unique values.
tech_support has 3 unique values.
streaming_tv has 3 unique values.
streaming_movies has 3 unique values.
paperless_billing has 2 unique values.
total_charges has 6531 unique values.
churn has 2 unique values.
internet_service_type has 3 unique values.
contract_type has 3 unique values.
payment_type has 4 unique values.
churn_month has 1 unique values.
paperless_billing.1 has 2 unique values.
gender.1 has 2 unique values.
partner.1 has 2 unique values.
dependents.1 has 2 unique values.
total_charges.1 has 6531 unique values.
signup_date has 73 unique values.
phone_service.1 has 2 unique values.
multiple_lines.1 has 3 unique values.
online_security.1 has 3 unique values.
online_backup.

In [10]:
ee.nunique_column_qty(df)

payment_type_id has 4 unique values.
contract_type_id has 3 unique values.
internet_service_type_id has 3 unique values.
senior_citizen has 2 unique values.
tenure has 73 unique values.
monthly_charges has 1585 unique values.
contract_type_id.1 has 3 unique values.
senior_citizen.1 has 2 unique values.
payment_type_id.1 has 4 unique values.
monthly_charges.1 has 1585 unique values.
internet_service_type_id.1 has 3 unique values.


In [11]:
df.multiple_lines.unique()

array(['No', 'Yes', 'No phone service'], dtype=object)

In [12]:
df.online_security.unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [13]:
df.online_backup.unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [14]:
df.device_protection.unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [15]:
df.tech_support.unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [16]:
df.streaming_tv.unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [17]:
df.streaming_movies.unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [18]:
df.internet_service_type.unique()

array(['DSL', 'Fiber optic', 'None'], dtype=object)

In [19]:
df.contract_type.unique()

array(['One year', 'Month-to-month', 'Two year'], dtype=object)

In [20]:
df.payment_type.unique()

array(['Mailed check', 'Electronic check', 'Credit card (automatic)',
       'Bank transfer (automatic)'], dtype=object)

# PREPARE

### PREPARE To-Do List

* DROP 
    * customer_id
    * payment_type_id
    * contract_type_id
    * internet_service_type_id
    * churn_month
    * paperless_billing.1
    * gender.1 
    * partner.1 
    * dependents.1 
    * total_charges.1 
    * phone_service.1 
    * multiple_lines.1 
    * online_security.1
    * online_backup.1 
    * device_protection.1
    * tech_support.1
    * streaming_tv.1
    * streaming_movies.1
    * contract_type_id.1 
    * senior_citizen.1 
    * payment_type_id.1 
    * monthly_charges.1 
    * internet_service_type_id.1 
    
* .rename
    * NONE

* .replace
    * churn_month has nulls; replace nulls with 'active' DROP (ALL DATES DUPLICATED).
    * multiple_lines has 3 unique values. REPLACE 'No phone service' with 'No'
    * online_security has 3 unique values. REPLACE 'No internet service' with 'No'
    * online_backup has 3 unique values. REPLACE 'No internet service' with 'No'
    * device_protection has 3 unique values. REPLACE 'No internet service' with 'No'
    * tech_support has 3 unique values. REPLACE 'No internet service' with 'No'
    * streaming_tv has 3 unique values. REPLACE 'No internet service' with 'No'
    * streaming_movies has 3 unique values. REPLACE 'No internet service' with 'No'

* reformat
    * signup_date 7043 non-null   object. REFORMAT OBJECT astype DATETIME
    * total_charges contains empty " " values. REPLACE empty " " values
    * total_charges 7043 non-null   object. REFORMAT OBJECT astype FLOAT

* encode.map
    * gender has 2 unique values.
    * partner has 2 unique values.
    * dependents has 2 unique values.
    * phone_service has 2 unique values.
    * paperless_billing has 2 unique values.
    * churn has 2 unique values.
    * senior_citizen has 2 unique values.
    * multiple_lines will have 2 unique values. REPLACED 'No phone service' with 'No'
    * online_security will have 2 unique values. REPLACED 'No internet service' with 'No'
    * online_backup will have 2 unique values. REPLACED 'No internet service' with 'No'
    * device_protection will have 2 unique values. REPLACED 'No internet service' with 'No'
    * tech_support will have 2 unique values. REPLACED 'No internet service' with 'No'
    * streaming_tv will have 2 unique values. REPLACED 'No internet service' with 'No'
    * streaming_movies will have 2 unique values. REPLACED 'No internet service' with 'No'

* melt
    * NONE

* pivot
    * internet_service_type has 3 unique values.
    * contract_type has 3 unique values.
    * payment_type has 4 unique values.

* Impute
    * NONE

# DROP
    * customer_id
    * payment_type_id
    * contract_type_id
    * internet_service_type_id
    * churn_month
    * paperless_billing.1
    * gender.1 
    * partner.1 
    * dependents.1 
    * total_charges.1 
    * phone_service.1 
    * multiple_lines.1 
    * online_security.1
    * online_backup.1 
    * device_protection.1
    * tech_support.1
    * streaming_tv.1
    * streaming_movies.1
    * contract_type_id.1 
    * senior_citizen.1 
    * payment_type_id.1 
    * monthly_charges.1 
    * internet_service_type_id.1 

In [21]:
df = df.drop(columns=['customer_id',
             'payment_type_id',
             'contract_type_id',
             'internet_service_type_id',
             'churn_month',
             'paperless_billing.1',
             'gender.1', 
             'partner.1', 
             'dependents.1', 
             'total_charges.1', 
             'phone_service.1', 
             'multiple_lines.1', 
             'online_security.1',
             'online_backup.1', 
             'device_protection.1',
             'tech_support.1',
             'streaming_tv.1',
             'streaming_movies.1',
             'contract_type_id.1', 
             'senior_citizen.1', 
             'payment_type_id.1', 
             'monthly_charges.1', 
             'internet_service_type_id.1'
            ])

In [22]:
df.T

0                    1     \
gender                              Female                 Male   
senior_citizen                           0                    0   
partner                                Yes                   No   
dependents                             Yes                   No   
tenure                                   9                    9   
phone_service                          Yes                  Yes   
multiple_lines                          No                  Yes   
online_security                         No                   No   
online_backup                          Yes                   No   
device_protection                       No                   No   
tech_support                           Yes                   No   
streaming_tv                           Yes                   No   
streaming_movies                        No                  Yes   
paperless_billing                      Yes                   No   
monthly_charges                       65.6                 59.9   
total_charges                        593.3                542.4   
churn                                   No                   No   
internet_service_type                  DSL                  DSL   
contract_type                     One year       Month-to-month   
payment_type                  Mailed check         Mailed check   
signup_date            2021-04-21 18:07:34  2021-04-21 18:07:34   

                                      2                    3     \
gender                                Male                 Male   
senior_citizen                           0                    1   
partner                                 No                  Yes   
dependents                              No                   No   
tenure                                   4                   13   
phone_service                          Yes                  Yes   
multiple_lines                          No                   No   
online_security                         No                   No   
online_backup                           No                  Yes   
device_protection                      Yes                  Yes   
tech_support                            No                   No   
streaming_tv                            No                  Yes   
streaming_movies                        No                  Yes   
paperless_billing                      Yes                  Yes   
monthly_charges                       73.9                 98.0   
total_charges                       280.85              1237.85   
churn                                  Yes                  Yes   
internet_service_type          Fiber optic          Fiber optic   
contract_type               Month-to-month       Month-to-month   
payment_type              Electronic check     Electronic check   
signup_date            2021-09-21 18:07:34  2020-12-21 18:07:34   

                                      4                        5     \
gender                              Female                   Female   
senior_citizen                           1                        0   
partner                                Yes                       No   
dependents                              No                      Yes   
tenure                                   3                        9   
phone_service                          Yes                      Yes   
multiple_lines                          No                       No   
online_security                         No                       No   
online_backup                           No                       No   
device_protection                       No                       No   
tech_support                           Yes                      Yes   
streaming_tv                           Yes                      Yes   
streaming_movies                        No                      Yes   
paperless_billing                      Yes                      Yes   
monthly_charges        

# REPLACE
    * churn_month has nulls; replace nulls with 'active' DROP (ALL DATES DUPLICATED).
    * multiple_lines has 3 unique values. REPLACE 'No phone service' with 'No'
    * online_security has 3 unique values. REPLACE 'No internet service' with 'No'
    * online_backup has 3 unique values. REPLACE 'No internet service' with 'No'
    * device_protection has 3 unique values. REPLACE 'No internet service' with 'No'
    * tech_support has 3 unique values. REPLACE 'No internet service' with 'No'
    * streaming_tv has 3 unique values. REPLACE 'No internet service' with 'No'
    * streaming_movies has 3 unique values. REPLACE 'No internet service' with 'No'

In [23]:
df.multiple_lines = df.multiple_lines.replace('No phone service', 'No')

df.online_security = df.online_security.replace('No internet service', 'No')
df.online_backup = df.online_backup.replace('No internet service', 'No')
df.device_protection = df.device_protection.replace('No internet service', 'No')
df.tech_support = df.tech_support.replace('No internet service', 'No')
df.streaming_tv = df.streaming_tv.replace('No internet service', 'No')
df.streaming_movies = df.streaming_movies.replace('No internet service', 'No')

In [24]:
df.multiple_lines.unique() 

array(['No', 'Yes'], dtype=object)

In [25]:
df.online_security.unique()  

array(['No', 'Yes'], dtype=object)

In [26]:
df.online_backup.unique()  

array(['Yes', 'No'], dtype=object)

In [27]:
df.device_protection.unique()

array(['No', 'Yes'], dtype=object)

In [28]:
df.tech_support.unique()

array(['Yes', 'No'], dtype=object)

In [29]:
df.streaming_tv.unique()

array(['Yes', 'No'], dtype=object)

In [30]:
df.streaming_movies.unique()

array(['No', 'Yes'], dtype=object)

# ENCODE.MAP
    * gender has 2 unique values.
    * partner has 2 unique values.
    * dependents has 2 unique values.
    * phone_service has 2 unique values.
    * paperless_billing has 2 unique values.
    * churn has 2 unique values.
    * multiple_lines will have 2 unique values. REPLACED 'No phone service' with 'No'
    * online_security will have 2 unique values. REPLACED 'No internet service' with 'No'
    * online_backup will have 2 unique values. REPLACED 'No internet service' with 'No'
    * device_protection will have 2 unique values. REPLACED 'No internet service' with 'No'
    * tech_support will have 2 unique values. REPLACED 'No internet service' with 'No'
    * streaming_tv will have 2 unique values. REPLACED 'No internet service' with 'No'
    * streaming_movies will have 2 unique values. REPLACED 'No internet service' with 'No'

In [31]:
df.gender.unique()

array(['Female', 'Male'], dtype=object)

In [32]:
df.partner.unique()

array(['Yes', 'No'], dtype=object)

In [33]:
df.dependents.unique()

array(['Yes', 'No'], dtype=object)

In [34]:
df.phone_service.unique()

array(['Yes', 'No'], dtype=object)

In [35]:
df.paperless_billing.unique()

array(['Yes', 'No'], dtype=object)

In [36]:
df.churn.unique()

array(['No', 'Yes'], dtype=object)

In [37]:
df.senior_citizen.unique()

array([0, 1])

In [38]:
df.multiple_lines.unique()

array(['No', 'Yes'], dtype=object)

In [39]:
df.online_security.unique()

array(['No', 'Yes'], dtype=object)

In [40]:
df.online_backup.unique()

array(['Yes', 'No'], dtype=object)

In [41]:
df.device_protection.unique()

array(['No', 'Yes'], dtype=object)

In [42]:
df.tech_support.unique()

array(['Yes', 'No'], dtype=object)

In [43]:
df.streaming_tv.unique()

array(['Yes', 'No'], dtype=object)

In [44]:
df.streaming_movies.unique()

array(['No', 'Yes'], dtype=object)

In [45]:
encoded_df = pd.DataFrame()
encoded_df['gender_encoded'] = df.gender.map({'Male': 1, 'Female': 0})
encoded_df['partner_encoded'] = df.partner.map({'Yes': 1, 'No': 0})
encoded_df['dependents_encoded'] = df.dependents.map({'Yes': 1, 'No': 0})
encoded_df['phone_service_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df['paperless_billing_encoded'] = df.paperless_billing.map({'Yes': 1, 'No': 0})
encoded_df['churn_encoded'] = df.churn.map({'Yes': 1, 'No': 0})
encoded_df['multiple_lines_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df['online_security_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df['online_backup_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df['device_protection_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df['tech_support_encoded'] = df.churn.map({'Yes': 1, 'No': 0})
encoded_df['streaming_tv_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df['streaming_movies_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df.T

0     1     2     3     4     5     6     7     \
gender_encoded                0     1     1     1     0     0     0     1   
partner_encoded               1     0     0     1     1     0     1     1   
dependents_encoded            1     0     0     0     0     1     0     0   
phone_service_encoded         1     1     1     1     1     1     1     1   
paperless_billing_encoded     1     0     1     1     1     1     1     1   
churn_encoded                 0     0     1     1     1     0     0     0   
multiple_lines_encoded        1     1     1     1     1     1     1     1   
online_security_encoded       1     1     1     1     1     1     1     1   
online_backup_encoded         1     1     1     1     1     1     1     1   
device_protection_encoded     1     1     1     1     1     1     1     1   
tech_support_encoded          0     0     1     1     1     0     0     0   
streaming_tv_encoded          1     1     1     1     1     1     1     1   
streaming_movies_encoded      1     1     1     1     1     1     1     1   

                           8     9     ...  7033  7034  7035  7036  7037  \
gender_encoded                0     0  ...     0     1     0     1     1   
partner_encoded               0     1  ...     1     0     0     1     0   
dependents_encoded            0     1  ...     0     0     0     1     0   
phone_service_encoded         1     1  ...     1     1     1     1     1   
paperless_billing_encoded     1     1  ...     1     1     0     0     0   
churn_encoded                 0     0  ...     0     1     0     0     0   
multiple_lines_encoded        1     1  ...     1     1     1     1     1   
online_security_encoded       1     1  ...     1     1     1     1     1   
online_backup_encoded         1     1  ...     1     1     1     1     1   
device_protection_encoded     1     1  ...     1     1     1     1     1   
tech_support_encoded          0     0  ...     0     1     0     0     0   
streaming_tv_encoded          1     1  ...     1     1     1     1     1   
streaming_movies_encoded      1     1  ...     1     1     1     1     1   

                           7038  7039  7040  7041  7042  
gender_encoded                0     0     0     1     1  
partner_encoded               0     0     0     0     1  
dependents_encoded            0     0     0     0     1  
phone_service_encoded         1     1     1     1     0  
paperless_billing_encoded     1     1     0     1     0  
churn_encoded                 0     1     1     0     0  
multiple_lines_encoded        1     1     1     1     0  
online_security_encoded       1     1     1     1     0  
online_backup_encoded         1     1     1     1     0  
device_protection_encoded     1     1     1     1     0  
tech_support_encoded          0     1     1     0     0  
streaming_tv_encoded          1     1     1     1     0  
streaming_movies_encoded      1     1     1     1     0  

[13 rows x 7043 columns]

In [46]:
ee.nunique_column_all(encoded_df)

1    3555
0    3488
Name: gender_encoded, dtype: int64

0    3641
1    3402
Name: partner_encoded, dtype: int64

0    4933
1    2110
Name: dependents_encoded, dtype: int64

1    6361
0     682
Name: phone_service_encoded, dtype: int64

1    4171
0    2872
Name: paperless_billing_encoded, dtype: int64

0    5174
1    1869
Name: churn_encoded, dtype: int64

1    6361
0     682
Name: multiple_lines_encoded, dtype: int64

1    6361
0     682
Name: online_security_encoded, dtype: int64

1    6361
0     682
Name: online_backup_encoded, dtype: int64

1    6361
0     682
Name: device_protection_encoded, dtype: int64

0    5174
1    1869
Name: tech_support_encoded, dtype: int64

1    6361
0     682
Name: streaming_tv_encoded, dtype: int64

1    6361
0     682
Name: streaming_movies_encoded, dtype: int64



# PIVOT
    * internet_service_type has 3 unique values.
    * contract_type has 3 unique values.
    * payment_type has 4 unique values.

In [47]:
# Use pandas dummies to pivot features with more than two string values
# into multiple columns with binary int values that can be read as boolean
# drop_first = False in draft for human readability; Final will have it set to True.

dummy_df = pd.get_dummies(data=df[['internet_service_type', 'contract_type', 'payment_type']], drop_first=False)
dummy_df.T

0     1     2     3     4     5     \
internet_service_type_DSL                  1     1     0     0     0     1   
internet_service_type_Fiber optic          0     0     1     1     1     0   
internet_service_type_None                 0     0     0     0     0     0   
contract_type_Month-to-month               0     1     1     1     1     1   
contract_type_One year                     1     0     0     0     0     0   
contract_type_Two year                     0     0     0     0     0     0   
payment_type_Bank transfer (automatic)     0     0     0     0     0     0   
payment_type_Credit card (automatic)       0     0     0     0     0     1   
payment_type_Electronic check              0     0     1     1     0     0   
payment_type_Mailed check                  1     1     0     0     1     0   

                                        6     7     8     9     ...  7033  \
internet_service_type_DSL                  0     0     1     1  ...     0   
internet_service_type_Fiber optic          1     1     0     0  ...     0   
internet_service_type_None                 0     0     0     0  ...     1   
contract_type_Month-to-month               0     0     1     0  ...     0   
contract_type_One year                     0     0     0     0  ...     0   
contract_type_Two year                     1     1     0     1  ...     1   
payment_type_Bank transfer (automatic)     1     0     0     0  ...     0   
payment_type_Credit card (automatic)       0     1     0     0  ...     1   
payment_type_Electronic check              0     0     1     0  ...     0   
payment_type_Mailed check                  0     0     0     1  ...     0   

                                        7034  7035  7036  7037  7038  7039  \
internet_service_type_DSL                  0     1     0     0     0     0   
internet_service_type_Fiber optic          1     0     0     0     1     1   
internet_service_type_None                 0     0     1     1     0     0   
contract_type_Month-to-month               1     1     0     1     0     1   
contract_type_One year                     0     0     0     0     1     0   
contract_type_Two year                     0     0     1     0     0     0   
payment_type_Bank transfer (automatic)     0     1     0     0     0     0   
payment_type_Credit card (automatic)       0     0     1     0     0     0   
payment_type_Electronic check              1     0     0     0     0     0   
payment_type_Mailed check                  0     0     0     1     1     1   

                                        7040  7041  7042  
internet_service_type_DSL                  0     1     1  
internet_service_type_Fiber optic          0     0     0  
internet_service_type_None                 1     0     0  
contract_type_Month-to-month               1     1     0  
contract_type_One year                     0     0     0  
contract_type_Two year                     0     0     1  
payment_type_Bank transfer (automatic)     1     0     0  
payment_type_Credit card (automatic)       0     0     0  
payment_type_Electronic check              0     0     1  
payment_type_Mailed check                  0     1     0  

[10 rows x 7043 columns]

In [48]:
ee.nunique_column_all(dummy_df)

0    4622
1    2421
Name: internet_service_type_DSL, dtype: int64

0    3947
1    3096
Name: internet_service_type_Fiber optic, dtype: int64

0    5517
1    1526
Name: internet_service_type_None, dtype: int64

1    3875
0    3168
Name: contract_type_Month-to-month, dtype: int64

0    5570
1    1473
Name: contract_type_One year, dtype: int64

0    5348
1    1695
Name: contract_type_Two year, dtype: int64

0    5499
1    1544
Name: payment_type_Bank transfer (automatic), dtype: int64

0    5521
1    1522
Name: payment_type_Credit card (automatic), dtype: int64

0    4678
1    2365
Name: payment_type_Electronic check, dtype: int64

0    5431
1    1612
Name: payment_type_Mailed check, dtype: int64



In [49]:
df = pd.concat([df, encoded_df, dummy_df], axis=1)

In [50]:
df.T

0     \
gender                                               Female   
senior_citizen                                            0   
partner                                                 Yes   
dependents                                              Yes   
tenure                                                    9   
phone_service                                           Yes   
multiple_lines                                           No   
online_security                                          No   
online_backup                                           Yes   
device_protection                                        No   
tech_support                                            Yes   
streaming_tv                                            Yes   
streaming_movies                                         No   
paperless_billing                                       Yes   
monthly_charges                                        65.6   
total_charges                                         593.3   
churn                                                    No   
internet_service_type                                   DSL   
contract_type                                      One year   
payment_type                                   Mailed check   
signup_date                             2021-04-21 18:07:34   
gender_encoded                                            0   
partner_encoded                                           1   
dependents_encoded                                        1   
phone_service_encoded                                     1   
paperless_billing_encoded                                 1   
churn_encoded                                             0   
multiple_lines_encoded                                    1   
online_security_encoded                                   1   
online_backup_encoded                                     1   
device_protection_encoded                                 1   
tech_support_encoded                                      0   
streaming_tv_encoded                                      1   
streaming_movies_encoded                                  1   
internet_service_type_DSL                                 1   
internet_service_type_Fiber optic                         0   
internet_service_type_None                                0   
contract_type_Month-to-month                              0   
contract_type_One year                                    1   
contract_type_Two year                                    0   
payment_type_Bank transfer (automatic)                    0   
payment_type_Credit card (automatic)                      0   
payment_type_Electronic check                             0   
payment_type_Mailed check                                 1   

                                                       1     \
gender                                                 Male   
senior_citizen                                            0   
partner                                                  No   
dependents                                               No   
tenure                                                    9   
phone_service                                           Yes   
multiple_lines                                          Yes   
online_security                                          No   
online_backup                                            No   
device_protection                                        No   
tech_support                                             No   
streaming_tv                                             No   
streaming_movies                                        Yes   
paperless_billing                                        No   
monthly_charges                                        59.9   
total_charges                                         542.4   
churn                                                    No   
internet_service_type                                   DSL   
contract_type         

In [51]:
# Drop encoded.map columns
df = df.drop(columns=['gender',
                      'partner',
                      'dependents',
                      'phone_service',
                      'paperless_billing',
                      'churn',
                      'multiple_lines',
                      'online_security',
                      'online_backup',
                      'device_protection',
                      'tech_support',
                      'streaming_tv',
                      'streaming_movies'])

# Drop pivot columns
df = df.drop(columns=['internet_service_type', 
                      'contract_type', 
                      'payment_type'])

In [52]:
ee.nunique_column_all(df)

0    5901
1    1142
Name: senior_citizen, dtype: int64

1     613
72    362
2     238
3     200
4     176
     ... 
28     57
39     56
44     51
36     50
0      11
Name: tenure, Length: 73, dtype: int64

20.05    61
19.85    45
19.95    44
19.90    44
20.00    43
         ..
97.45     1
92.10     1
51.60     1
78.25     1
91.75     1
Name: monthly_charges, Length: 1585, dtype: int64

20.2       11
           11
19.75       9
19.9        8
19.65       8
           ..
985.05      1
1396        1
435         1
3720.35     1
3707.6      1
Name: total_charges, Length: 6531, dtype: int64

2021-12-21 18:07:34    613
2016-01-21 18:07:34    362
2021-11-21 18:07:34    238
2021-10-21 18:07:34    200
2021-09-21 18:07:34    176
                      ... 
2019-09-21 18:07:34     57
2018-10-21 18:07:34     56
2018-05-21 18:07:34     51
2019-01-21 18:07:34     50
2022-01-21 18:07:34     11
Name: signup_date, Length: 73, dtype: int64

1    3555
0    3488
Name: gender_encoded, dtype: int64

0    3641


In [53]:
df.T

0     \
senior_citizen                                            0   
tenure                                                    9   
monthly_charges                                        65.6   
total_charges                                         593.3   
signup_date                             2021-04-21 18:07:34   
gender_encoded                                            0   
partner_encoded                                           1   
dependents_encoded                                        1   
phone_service_encoded                                     1   
paperless_billing_encoded                                 1   
churn_encoded                                             0   
multiple_lines_encoded                                    1   
online_security_encoded                                   1   
online_backup_encoded                                     1   
device_protection_encoded                                 1   
tech_support_encoded                                      0   
streaming_tv_encoded                                      1   
streaming_movies_encoded                                  1   
internet_service_type_DSL                                 1   
internet_service_type_Fiber optic                         0   
internet_service_type_None                                0   
contract_type_Month-to-month                              0   
contract_type_One year                                    1   
contract_type_Two year                                    0   
payment_type_Bank transfer (automatic)                    0   
payment_type_Credit card (automatic)                      0   
payment_type_Electronic check                             0   
payment_type_Mailed check                                 1   

                                                       1     \
senior_citizen                                            0   
tenure                                                    9   
monthly_charges                                        59.9   
total_charges                                         542.4   
signup_date                             2021-04-21 18:07:34   
gender_encoded                                            1   
partner_encoded                                           0   
dependents_encoded                                        0   
phone_service_encoded                                     1   
paperless_billing_encoded                                 0   
churn_encoded                                             0   
multiple_lines_encoded                                    1   
online_security_encoded                                   1   
online_backup_encoded                                     1   
device_protection_encoded                                 1   
tech_support_encoded                                      0   
streaming_tv_encoded                                      1   
streaming_movies_encoded                                  1   
internet_service_type_DSL                                 1   
internet_service_type_Fiber optic                         0   
internet_service_type_None                                0   
contract_type_Month-to-month                              1   
contract_type_One year                                    0   
contract_type_Two year                                    0   
payment_type_Bank transfer (automatic)                    0   
payment_type_Credit card (automatic)                      0   
payment_type_Electronic check                             0   
payment_type_Mailed check                                 1   

                                                       2     \
senior_citizen                                            0   
tenure                                                    4   
monthly_charges                                        73.9   
total_charges                                        280.85   
signup_date                             2021-09-21 18:07:34   
gender_encoded       

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 28 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   senior_citizen                          7043 non-null   int64  
 1   tenure                                  7043 non-null   int64  
 2   monthly_charges                         7043 non-null   float64
 3   total_charges                           7043 non-null   object 
 4   signup_date                             7043 non-null   object 
 5   gender_encoded                          7043 non-null   int64  
 6   partner_encoded                         7043 non-null   int64  
 7   dependents_encoded                      7043 non-null   int64  
 8   phone_service_encoded                   7043 non-null   int64  
 9   paperless_billing_encoded               7043 non-null   int64  
 10  churn_encoded                           7043 non-null   int6

In [55]:
df.total_charges.info()

<class 'pandas.core.series.Series'>
Int64Index: 7043 entries, 0 to 7042
Series name: total_charges
Non-Null Count  Dtype 
--------------  ----- 
7043 non-null   object
dtypes: object(1)
memory usage: 368.1+ KB


In [56]:
df.total_charges = df.total_charges.replace(' ', '0')

In [57]:
#df['total_charges'] = df['total_charges'].astype(float)
#df.total_charges = df.total_charges.astype(float)
df['total_charges'] = pd.to_numeric(df['total_charges'])

In [58]:
df.total_charges.info()

<class 'pandas.core.series.Series'>
Int64Index: 7043 entries, 0 to 7042
Series name: total_charges
Non-Null Count  Dtype  
--------------  -----  
7043 non-null   float64
dtypes: float64(1)
memory usage: 368.1 KB


In [59]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 28 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   senior_citizen                          7043 non-null   int64  
 1   tenure                                  7043 non-null   int64  
 2   monthly_charges                         7043 non-null   float64
 3   total_charges                           7043 non-null   float64
 4   signup_date                             7043 non-null   object 
 5   gender_encoded                          7043 non-null   int64  
 6   partner_encoded                         7043 non-null   int64  
 7   dependents_encoded                      7043 non-null   int64  
 8   phone_service_encoded                   7043 non-null   int64  
 9   paperless_billing_encoded               7043 non-null   int64  
 10  churn_encoded                           7043 non-null   int6

In [60]:
df["signup_date"] = pd.to_datetime(df["signup_date"])

In [61]:
df.signup_date.info()

<class 'pandas.core.series.Series'>
Int64Index: 7043 entries, 0 to 7042
Series name: signup_date
Non-Null Count  Dtype         
--------------  -----         
7043 non-null   datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 368.1 KB


In [62]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 28 columns):
 #   Column                                  Non-Null Count  Dtype         
---  ------                                  --------------  -----         
 0   senior_citizen                          7043 non-null   int64         
 1   tenure                                  7043 non-null   int64         
 2   monthly_charges                         7043 non-null   float64       
 3   total_charges                           7043 non-null   float64       
 4   signup_date                             7043 non-null   datetime64[ns]
 5   gender_encoded                          7043 non-null   int64         
 6   partner_encoded                         7043 non-null   int64         
 7   dependents_encoded                      7043 non-null   int64         
 8   phone_service_encoded                   7043 non-null   int64         
 9   paperless_billing_encoded               7043 non-nul

# Data Dictionary

| Feature | Definition |
|:--------|:-----------|
|Rated| True or False, The game's result is reflected in each player's rating|
|Winning Pieces| The color of pieces the winning player was moving|
|White Rating| Rating of the player moving the white pieces using the Glicko-2 rating method for games played on Lichess|
|Black Rating| Rating of the player moving the white pieces using the Glicko-2 rating method for games played on Lichess|
|Rating Difference| The difference in rating between the players in the game|
|Game Rating| The average rating of the two players in the game|
|Lower Rated White| True or False, The lower rated player is moving the white pieces|
|Opening Name| The name of the opening played in the game|
|Time Control Group| The amount of time allotted to each player to make their moves, **Standard** (60 min or more), **Rapid** (30 - 15 min), **Blitz** (5 - 3 min), or **Bullet** (2 or less), **Other** (any other time limit)|
|Upset (Target)| True or False, The lower rated player won the game|
|Additional Features|Encoded and values for categorical data and scaled versions continuous data|

In [63]:
df.columns

Index(['senior_citizen', 'tenure', 'monthly_charges', 'total_charges',
       'signup_date', 'gender_encoded', 'partner_encoded',
       'dependents_encoded', 'phone_service_encoded',
       'paperless_billing_encoded', 'churn_encoded', 'multiple_lines_encoded',
       'online_security_encoded', 'online_backup_encoded',
       'device_protection_encoded', 'tech_support_encoded',
       'streaming_tv_encoded', 'streaming_movies_encoded',
       'internet_service_type_DSL', 'internet_service_type_Fiber optic',
       'internet_service_type_None', 'contract_type_Month-to-month',
       'contract_type_One year', 'contract_type_Two year',
       'payment_type_Bank transfer (automatic)',
       'payment_type_Credit card (automatic)', 'payment_type_Electronic check',
       'payment_type_Mailed check'],
      dtype='object')

# SPLIT

In [64]:
# use predefined function to split data for train, validate, and test
target = 'churn_encoded'
train_df, validate_df, test_df = w.split(df, target)

_______________________________________________________________
|                              DF                             |
|-------------------|-------------------|---------------------|
|       Train       |       Validate    |          Test       |
|-------------------|-------------------|-----------|---------|
| x_train | y_train |   x_val  |  y_val |   x_test  |  y_test |
|-------------------------------------------------------------|

* 1. tree_1 = DecisionTreeClassifier(max_depth = 5)
* 2. tree_1.fit(x_train, y_train)
* 3. predictions = tree_1.predict(x_train)
* 4. pd.crosstab(y_train, predictions)
* 5. val_predictions = tree_1.predict(x_val)
* 6. pd.crosstab(y_val, val_predictions)

Prepared df: (7043, 28)

Train (train_df): (4225, 28)
Validate (validate_df): (1409, 28)
Test (test_df): (1409, 28)


# MODEL

In [65]:
train_df.T

0     \
senior_citizen                                            1   
tenure                                                   44   
monthly_charges                                       109.8   
total_charges                                       4860.35   
signup_date                             2018-05-21 18:07:34   
gender_encoded                                            0   
partner_encoded                                           0   
dependents_encoded                                        0   
phone_service_encoded                                     1   
paperless_billing_encoded                                 1   
churn_encoded                                             0   
multiple_lines_encoded                                    1   
online_security_encoded                                   1   
online_backup_encoded                                     1   
device_protection_encoded                                 1   
tech_support_encoded                                      0   
streaming_tv_encoded                                      1   
streaming_movies_encoded                                  1   
internet_service_type_DSL                                 0   
internet_service_type_Fiber optic                         1   
internet_service_type_None                                0   
contract_type_Month-to-month                              0   
contract_type_One year                                    1   
contract_type_Two year                                    0   
payment_type_Bank transfer (automatic)                    0   
payment_type_Credit card (automatic)                      1   
payment_type_Electronic check                             0   
payment_type_Mailed check                                 0   

                                                       1     \
senior_citizen                                            0   
tenure                                                   14   
monthly_charges                                        55.7   
total_charges                                        795.15   
signup_date                             2020-11-21 18:07:34   
gender_encoded                                            0   
partner_encoded                                           0   
dependents_encoded                                        0   
phone_service_encoded                                     1   
paperless_billing_encoded                                 0   
churn_encoded                                             0   
multiple_lines_encoded                                    1   
online_security_encoded                                   1   
online_backup_encoded                                     1   
device_protection_encoded                                 1   
tech_support_encoded                                      0   
streaming_tv_encoded                                      1   
streaming_movies_encoded                                  1   
internet_service_type_DSL                                 1   
internet_service_type_Fiber optic                         0   
internet_service_type_None                                0   
contract_type_Month-to-month                              0   
contract_type_One year                                    1   
contract_type_Two year                                    0   
payment_type_Bank transfer (automatic)                    0   
payment_type_Credit card (automatic)                      0   
payment_type_Electronic check                             0   
payment_type_Mailed check                                 1   

                                                       2     \
senior_citizen                                            0   
tenure                                                   14   
monthly_charges                                      100.55   
total_charges                                       1415.55   
signup_date                             2020-11-21 18:07:34   
gender_encoded       